pkg/loop/internal/chain_reader.go

```diff
@@ -0,0 +1,293 @@
+package internal
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"unicode"
+
+	"github.com/fxamacker/cbor/v2"
+	libocr "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	"google.golang.org/grpc/status"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/loop/internal/pb"
+	"github.com/smartcontractkit/chainlink-relay/pkg/types"
+)
+
+var _ types.ChainReader = (*chainReaderClient)(nil)
+
+type chainReaderClient struct {
+	*brokerExt
+	grpc pb.ChainReaderClient
+}
+
+// enum of all known encoding formats for versioned data
+const (
+	JSONEncodingVersion = iota
+	CBOREncodingVersion
+)
+
+// Version to be used for encoding ( version used for decoding is determined by data received )
+// These are separate constants in case we want to upgrade their data formats independently
+const ParamsCurrentEncodingVersion = CBOREncodingVersion
+const RetvalCurrentEncodingVersion = CBOREncodingVersion
+
+func encodeVersionedBytes(data any, version int32) (*pb.VersionedBytes, error) {
+	var bytes []byte
+	var err error
+
+	switch version {
+	case JSONEncodingVersion:
+		bytes, err = json.Marshal(data)
+	case CBOREncodingVersion:
+		enco := cbor.CoreDetEncOptions()
+		enco.Time = cbor.TimeRFC3339Nano
+		var enc cbor.EncMode
+		enc, err = enco.EncMode()
+		if err != nil {
+			return nil, err
+		}
+		bytes, err = enc.Marshal(data)
+	default:
+		return nil, fmt.Errorf("unsupported encoding version %d for data %v", version, data)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", types.InvalidTypeError{}, err)
```

Comment:
I'm worried that this shared error handling is brittle, especially given the asymmetry. Why don't we handle these directly in each case in the normal way? Could add context too.

Separately, is InvalidTypeError always associated with an error cause? If so, we can simplify these to InvalidTypeError{Err: err}. (Or ErrInvalidType)

Reply:
> InvalidTypeError always associated with an error cause? If so, we can simplify these to InvalidTypeError{Err: err}. (Or ErrInvalidType)

I'm pretty sure it's not, but I suppose we could always add a "None" or "Other" cause to the places where there is no reason.